### PR TITLE
Add Gradient for Reciprocal

### PR DESCRIPTION
**Pull request by baijumeswani**
Keras BN maps to ONNX primitive ops. And gradient building failed on `Reciprocal` for those primitive ops.

BN gradient in ORT cannot be substituted by primitive ops (since the stats need to be maintained). However, by the time I realized that, I had already implemented this gradient.

### CODE REVIEW

**Review by er3x3 — COMMENTED** [review]
Maybe we still need a C++ UT in orttraining/orttraining/test/gradient/gradient_ops_test.cc.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 1]
```diff
@@ -1994,5 +1994,14 @@ IMPLEMENT_GRADIENT_BUILDER(GetLSTMGradient) {
   return {NodeDef(OpDef{"LSTMGrad", kMSDomain, 1}, input_args, output_args, SrcNodeAttributes())};
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetReciprocalGradient) {
+  // y = 1 / x
+  // dy/dx = -1 / x^2
+  // dL/dx = dL/dy * dy/dx = dL/dy * (-1 / x^2)
+  return {NodeDef("Mul", {O(0), O(0)}, {IA("Square_O0")}),
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/core/graph/gradient_builder.cc:2001]: I used `Mul` because I thought that would be faster. I guess for such a simple op, it shouldn't really matter.

Will change to `Pow`.

**Review by pengwa — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -6147,3 +6147,33 @@ def forward(self, x):
...
+        loss.backward()
+        return prediction, loss
+
+    device = "cuda"
+    pt_model = ReciprocalModel().to(device)
+    ort_model = ORTModule(
+        copy.deepcopy(pt_model), DebugOptions(log_level=LogLevel.WARNING, save_onnx=True, onnx_prefix="reciprocal")
+    )
+
+    pt_x = torch.randn(3, 224, 224, requires_grad=True, device=device)
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6172]: looks this did not exclude 0 for input.

File: orttraining/orttraining/test/gradient/gradient_ops_test.cc [hunk 3]
```diff
@@ -3027,6 +3027,12 @@ TEST(GradientCheckerTest, PadAndUnflattenGrad) {
 }
 #endif
 
+TEST(GradientCheckerTest, ReciprocalGrad) {
+  // Avoid division by 0 by using the transformer.
+  std::function<float(float)> transformer = [](float x) { return x > 0 ? x + 0.2f : x - 0.2f; };
```
Comment (pengwa) [inline orttraining/orttraining/test/gradient/gradient_ops_test.cc:3032]: it is still possible to generate 0 right? if x = 0.2 or -0.2f

Comment (pengwa) [reply orttraining/orttraining/test/gradient/gradient_ops_test.cc:3032]: forget it.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -6147,3 +6147,33 @@ def forward(self, x):
...
+    def run_step(model, x):
+        prediction = model(x)
+        loss = prediction.sum()
+        loss.backward()
+        return prediction, loss
+
+    device = "cuda"
+    pt_model = ReciprocalModel().to(device)
+    ort_model = ORTModule(
+        copy.deepcopy(pt_model), DebugOptions(log_level=LogLevel.WARNING, save_onnx=True, onnx_prefix="reciprocal")
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6169]:
```suggestion
        copy.deepcopy(pt_model)
```

Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6169]: Thanks. I missed removing this after debugging.

**Review by pengwa — APPROVED** [review]

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (er3x3) [reply orttraining/orttraining/core/graph/gradient_builder.cc]: Is Pow really a good choice? See below code, since you are using int32 as T1 here, means it will always cast data to double for compute and cast back at the end.

```
// Capture permutations of int32/64/float/double
template <typename T, typename T1>
__device__ __inline__ T _Pow(T a, T1 b) {
  return static_cast<T>(pow(static_cast<double>(a), static_cast<double>(b)));
}

template <>
__device__ __inline__ float _Pow(float a, float b) { return powf(a, b); }

template <>
__device__ __inline__ double _Pow(double a, double b) { return pow(a, b); }

template <>
__device__ __inline__ half _Pow(half a, half b) { return half(powf((float)a, (float)b)); }
```

**Review by er3x3 — APPROVED** [review]

**Comment by v1w319** [comment]
Check the vC,:2881; then send command prompt to ci.yamva
